synapse/util/async_helpers.py

```diff
@@ -127,20 +136,13 @@ def observe(self) -> "defer.Deferred[_T]":
         """
         if not self._result:
             d: "defer.Deferred[_T]" = defer.Deferred()
-
-            def remove(r):
-                self._observers.discard(d)
-                return r
-
-            d.addBoth(remove)
-
-            self._observers.add(d)
+            self._observers.append(d)
             return d
         else:
             success, res = self._result
             return defer.succeed(res) if success else defer.fail(res)
 
-    def observers(self) -> "List[defer.Deferred[_T]]":
+    def observers(self) -> "Collection[defer.Deferred[_T]]":
```

Comment:
A bit surprised that we're leaking internal state here!

Could only see [one use](https://github.com/matrix-org/synapse/blob/c7a5e49664ab0bd18a57336e282fa6c3b9a17ca0/synapse/notifier.py#L162-L163) though. Probably fine as it is.

Reply:
yup. Thought about changing it to a method that just returns the number of observers. Ran out of enthusiasm.